--- __init__.py.orig	2010-02-11 19:53:01.000000000 +0000
+++ __init__.py	2010-02-11 19:53:57.000000000 +0000
@@ -60,6 +60,7 @@
 
 # try to use simplejson first, otherwise fallback to XML
 RESPONSE_FORMAT = 'JSON'
+RETRY_ATTEMPTS = 1
 try:
     import json as simplejson
 except ImportError:
@@ -100,8 +101,14 @@
 
 except ImportError:
     def urlread(url, data=None):
-        res = urllib2.urlopen(url, data=data)
-        return res.read()
+        global RETRY_ATTEMPTS
+        for attempt in xrange(0, RETRY_ATTEMPTS):
+            try:
+                res = urllib2.urlopen(url, data=data)
+                return res.read()
+            except Exception, e:
+                pass
+        raise Exception("Unable to contact facebook", e)
 
 __all__ = ['Facebook']
 
