--- __init__.py.orig	2009-04-20 00:18:04.000000000 +0100
+++ __init__.py	2009-05-06 13:41:27.000000000 +0100
@@ -433,12 +433,15 @@
             ('x', float, [('default', 50)]),
             ('y', float, [('default', 50)]),
             ('tags', str, ['optional']),
+            ('owner_uid', int, ['optional']),
         ],
 
         'createAlbum': [
             ('name', str, []),
             ('location', str, ['optional']),
             ('description', str, ['optional']),
+            ('visible', str, [('default', 'everyone')]),
+            ('uid', int, ['optional']),
         ],
 
         'get': [
